### ISSUE TITLE
[Question] Multiple queues

### ISSUE THREAD

**Issue by AustinJ235** [vulkano-rs/vulkano#1156]
I have two separate queues a graphics queue and a transfer queue. I want to upload/do work on the transfer queue at a low priority and then after completion switch to using the result on the graphics queue.

What is the proper way to wait on the operation on the transfer queue to complete and be ready for use on the graphics queue?

I think semaphores is the answer? But it is unclear how I am to wait on those.

**Comment by mitchmindtree** [comment]
It sounds like you're looking for one of the [`GpuFuture::then_signal_semaphore`](https://docs.rs/vulkano/0.11.1/vulkano/sync/trait.GpuFuture.html#method.then_signal_semaphore) or [`then_signal_semaphore_and_flush`](https://docs.rs/vulkano/0.11.1/vulkano/sync/trait.GpuFuture.html#method.then_signal_semaphore_and_flush) methods, but I'm not exactly sure of how to have one queue wait on another's semaphore. I'm guessing you might have to do something along the lines of this:

```rust
transfer_queue_work_future
    .then_signal_semaphore()
    .then_execute(graphics_queue, graphics_queue_work)
```

**Comment by AustinJ235** [comment]
I think doing it that way removes the benefit of using another queue as the main graphics queue will still be waiting on the result before it does any more work.

I need something like what the ``wait()`` method on fences I think. This might not be implemented?

My current intent is that on another cpu thread separate from the main cpu one, that work is done on the gpu on the transfer queue and then that cpu thread will wait for the work to be done. When it completes it will set some mutex somewhere or something of that like to the new value that then will be used when the main cpu thread is generating the command to run on the graphics queue.

**Comment by mitchmindtree** [comment]
> My current intent is that on another cpu thread separate from the main cpu one, that work is done on the gpu on the transfer queue and then that cpu thread will wait for the work to be done. When it completes it will set some mutex somewhere or something of that like to the new value that then will be used when the main cpu thread is generating the command to run on the graphics queue.

Ahh it sounds like for you probably want to use `GpuFuture::then_signal_fence_and_flush()` for your transfer queue and then `.wait()` on the `FenceSignalFuture` that is returned. Once `.wait()` returns, you know the work is done. Then you can signal your other cpu thread that that image is ready for use with your main graphics queue. Does that sound close to what you're after?

**Comment by AustinJ235** [comment]
Yes, but I think fences can return before the result can be used on another queue. I am not 100% sure on this though.

**Comment by mitchmindtree** [comment]
If you wait for `FenceSignalFuture::wait` to return it should definitely wait right until the end of that execution as this is the main purpose of fence (to synchronise CPU and GPU). I'm not sure if there is anything else required in preparing the image/data for processing on a different queue, though I would assume it would be fine if we've already `.wait()`ed for it to finish being processed on the previous queue?

**Comment by AustinJ235** [comment]
I am currently using the ``then_signal_fence_and_flush()`` followed by the ``wait()``, and this does work most of the time. Sometimes though when the graphics queue goes to use the result it will spit out an access error saying that it is already in use.

This could very well be a fault in my logic somewhere though as I use fences for a similar purpose elsewhere without issue. I'll play around with it a bit more.

**Comment by AustinJ235** [comment]
Seems the problem may have been related to been previously using the image in the graphics queue. I use two images, one active and one inactive. I transfer to the inactive one and then switch it to being active. I added an abstraction on top of ``ImageViewAccess`` to allow me to wait until the graphics queue has follow dropped the lease if you want to call it that. This seems to have solved the issue at least from the limited testing I have done.

I am guessing that even though I had marked the image as inactive, the main thread must have been out of date sometimes leading the graphics queue to try to use the image that the transfer queue was currently using.

I'll reopen the issue if I have any issues with this solution.
